src/mlpack/core/optimizers/cmaes/cmaes.hpp

```diff
@@ -0,0 +1,306 @@
+/**
+ * @file cmaes.hpp
+ * @author Kartik Nighania (GSoC 17 mentor Marcus Edel)
+ *
+ * Covariance Matrix Adaptation Evolution Strategy
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_CMAES_CMAES_HPP
+#define MLPACK_CORE_OPTIMIZERS_CMAES_CMAES_HPP
+
+#include "random.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+template<typename funcType>
+class CMAES
+{
+ public:
+CMAES(funcType& function, arma::mat& start, arma::mat& stdDivs,
+double iters = -1.0, double evalDiff = 1e-14);
+
+double Optimize(arma::mat& arr);
+
+size_t getDimension(void){ return N;}
+
+void getXBestEver(double *arr)
+{
+for (int i = 0; i < N; ++i) arr[i] = xBestEver[i];
+}
+
+void stopMaxFuncEvaluations(double evaluations)
+{
+stopMaxFunEvals = evaluations;
+}
+
+double getStopMaxFuncEvaluations(void)
+{
+  return  stopMaxFunEvals;
+}
+
+void stopMaxIterations(double iterations)
+{
+  stopMaxIter = iterations;
+}
+
+double getStopMaxIterations(void)
+{
+  return stopMaxIter;
+}
+
+void stopMinFuntionDifference(double difference)
+{
+  stopTolFun = difference;
+}
+
+double getStopMinFunctionDifference(void)
+{
+  return stopTolFun;
+}
+
+void stopMinFuntionHistoryDifference(double difference)
+{
+  stopTolFunHist = difference;
+}
+
+double getStopMinFunctionHistoryDifference(void)
+{
+  return stopTolFunHist;
+}
+
+void stopMinStepSize(double size)
+{
+  stopTolX = size;
+}
+
+double getStopMinStepSize(void)
+{
+  return stopTolX;
+}
+
+  //! other variable parameters
+
+void sampleSize(double l){lambda = l;}
+
+int getSampleSize(void){ return lambda; }
+
+void setMu(double ind){ mu = ind;}
+
+double getMu(void){ return mu;}
+
+void muEffective(double ind){ mueff = ind;}
+
+double getMuEffective(void){ return mueff;}
+
+double axisRatio() { return rgD.subvec(0, N) / rgD.subvec(0, N);}
+
+double evaluation(){ return countevals; }
+
+double fitness(){ return functionValues[(int)index[0]];}
+
+double fitnessBestEver(){ return xBestEver[N];}
+
+double generation(){ return gen;}
+
+double maxAxisLength(){ return sigma*std::sqrt(maxEW);}
+
+double minAxisLength(){ return sigma*std::sqrt(minEW); }
+
+double maxStdDev(){return sigma*std::sqrt(maxdiagC);}
+
+double minStdDev(){return sigma*std::sqrt(mindiagC);}
+
+void diagonalCovariance(double *arr)
```

Comment:
I see, I think we have to see which function is useful for a user an which could be removed. But we don't have do this now, but something like: ``void diagonalCovariance(double *arr)`` could be updated since, we use ``arma::mat``or ``arma:.vec``.